File: Napakalaki/src/NapakalakiGame/NumericBadConsequence.java
/**
 * NumericBadConsequence.java
 */

package NapakalakiGame;

import java.util.ArrayList;

/**
 * Representa un mal rollo numérico.
 */
public class NumericBadConsequence extends BadConsequence {
    
    private int nVisibleTreasures;
    private int nHiddenTreasures;
    
    final static int MAXTREASURES = 10;
    
    public NumericBadConsequence(String t, int l, int nVisible, int nHidden) {
        super(t, l);
        nVisibleTreasures = nVisible;
        nHiddenTreasures = nHidden;
    }
    
    public int getNVisibleTreasures() {
        return this.nVisibleTreasures;
    }
    
    public int getNHiddenTreasures() {
        return this.nHiddenTreasures;
    }
    
    @Override
    public boolean isEmpty() {
        return nVisibleTreasures == 0 && nHiddenTreasures == 0;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t) {
        if (nVisibleTreasures > 0) {
            nVisibleTreasures--;
        }
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t) {
        if (nHiddenTreasures > 0) {
            nHiddenTreasures--;
        }
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, 
                                                   ArrayList<Treasure> h) {
        // Nos quedamos con el mínimo
        int minVisible = Math.min(v.size(), nVisibleTreasures);
        int minHidden = Math.min(h.size(), nHiddenTreasures);
            
        return new NumericBadConsequence(getText(), getLevels(), minVisible, minHidden);
    }
    
    @Override
    public String toString() {
        return super.toString() + "\nTesoros visibles perdidos: " + 
               (nVisibleTreasures == MAXTREASURES ?
               "TODOS LOS TESOROS" : Integer.toString(nVisibleTreasures)) +
               "\nTesoros ocultos perdidos: " + (nHiddenTreasures == MAXTREASURES ?
               "TODOS LOS TESOROS" : Integer.toString(nHiddenTreasures));
    }
    
} 
        

File: Napakalaki/lib/numeric_bad_consequence.rb
# encoding: utf-8

require_relative 'bad_consequence'

module NapakalakiGame

  # class NumericBadConsequence: representa un mal rollo de número de tesoros
  class NumericBadConsequence < BadConsequence
    def initialize(t, l, v, h)
      super(t, l)
      @nVisibleTreasures = v
      @nHiddenTreasures = h
    end

    def getNVisibleTreasures
      @nVisibleTreasures
    end

    def getNHiddenTreasures
      @nHiddenTreasures
    end

    def isEmpty
      @nVisibleTreasures == 0 and @nHiddenTreasures == 0
    end

    def substractVisibleTreasure(t)
      if (@nVisibleTreasures > 0)
          @nVisibleTreasures -= 1
      end 
    end
    
    def substractHiddenTreasure(t)
      if @nHiddenTreasures > 0
          @nHiddenTreasures -= 1
      end
    end

    def adjustToFitTreasureLists(v,h)
      newNVisibleTreasures = [@nVisibleTreasures, v.count].min
      newNHiddenTreasures = [@nHiddenTreasures, h.count].min
      NumericBadConsequence.new(@text, @levels, newNVisibleTreasures, newNHiddenTreasures)
    end

    def to_s
      text = super
      text << "\n    * Tesoros visibles perdidos: " << if @nVisibleTreasures == @@MAXTREASURES
                                                   "TODOS LOS TESOROS"
                                                 else
                                                   "#{@nVisibleTreasures}"
                                                 end
      text << "\n    * Tesoros ocultos perdidos: " << if @nHiddenTreasures == @@MAXTREASURES
                                                  "TODOS LOS TESOROS"
                                                else
                                                  "#{@nHiddenTreasures}"
                                                end
    end

  end
  
end
